modules/fundamental/src/vulnerability/service/mod.rs

```diff
@@ -155,6 +162,141 @@ impl VulnerabilityService {
 
         Ok(result.rows_affected)
     }
+
+    pub async fn analyze_purls<C>(
+        &self,
+        purls: Vec<&str>,
+        connection: &C,
+    ) -> Result<HashMap<String, Vec<VulnerabilityDetails>>, Error>
+    where
+        C: ConnectionTrait + StreamTrait,
+    {
+        if purls.is_empty() {
+            return Ok(Default::default());
+        }
+
+        let queries: Vec<Result<String, Error>> = purls
+            .iter()
+            .map(|p| {
+                let purl = match Purl::from_str(p) {
+                    Ok(purl) => purl,
+                    Err(e) => return Err(Error::Purl(e)),
+                };
+
+                let version = match purl.version.as_ref() {
+                    Some(v) => v,
+                    None => return Err(Error::Purl(PurlErr::MissingVersion(p.to_string()))),
+                };
+
+                let ns_condition = match purl.namespace.clone() {
+                    Some(namespace) => {
+                        let sql = "base_purl.namespace = $1";
+                        Statement::from_sql_and_values(
+                            connection.get_database_backend(),
+                            sql,
+                            [namespace.into()],
+                        )
+                    }
+                    None => Statement::from_string(
+                        connection.get_database_backend(),
+                        "base_purl.namespace IS NULL",
+                    ),
+                };
+
+                let name = purl.name;
+                let ty = purl.ty;
+                let values = [
+                    p.to_string().into(),
+                    name.into(),
+                    ty.into(),
+                    version.to_string().into(),
+                ];
```

Comment:
These extra vars aren't needed
```suggestion
                let values = [
                    p.to_string().into(),
                    purl.name.into(),
                    purl.ty.into(),
                    purl.version.into(),
                ];
```
You also shouldn't have to use `p` here. You should be able to use `purl.to_string().into()` but for a bug that's fixed by #1460 once someone approves it.

Reply:
You should be able to do anything i can do. Where did you move `purl.ty` and why? could you have used a reference instead?

Reply:
If you're still assigning`ty = purl.ty` above that `values` assignment, then yeah, that'll move it and cause that error. But if you delete those ty/name lets, you should be good.

Reply:
Noob's mistake. Apologies.